cpp/include/raft/spatial/knn/detail/ann_quantized.cuh

```diff
@@ -0,0 +1,305 @@
+#include <rmm/mr/device/managed_memory_resource.hpp>
+
+#include <faiss/gpu/GpuDistance.h>
+#include <faiss/gpu/GpuIndexFlat.h>
+#include <faiss/gpu/GpuIndexIVFFlat.h>
+#include <faiss/gpu/GpuIndexIVFPQ.h>
+#include <faiss/gpu/GpuIndexIVFScalarQuantizer.h>
+#include <faiss/gpu/GpuResources.h>
+#include <faiss/gpu/utils/Limits.cuh>
+#include <faiss/gpu/utils/Select.cuh>
+#include <faiss/gpu/utils/Tensor.cuh>
+#include <faiss/utils/Heap.h>
+
+#include <thrust/iterator/transform_iterator.h>
+
+#include <raft/distance/distance_type.hpp>
+
+#include "ann_ivf_flat.cuh"
+
+#include <iostream>
+#include <set>
+
+#define IVF_FAISS 0
+
+namespace raft {
+namespace spatial {
+namespace knn {
+namespace detail {
+
+inline faiss::ScalarQuantizer::QuantizerType build_faiss_qtype(QuantizerType qtype)
+{
+  switch (qtype) {
+    case QuantizerType::QT_8bit: return faiss::ScalarQuantizer::QuantizerType::QT_8bit;
+    case QuantizerType::QT_8bit_uniform:
+      return faiss::ScalarQuantizer::QuantizerType::QT_8bit_uniform;
+    case QuantizerType::QT_4bit_uniform:
+      return faiss::ScalarQuantizer::QuantizerType::QT_4bit_uniform;
+    case QuantizerType::QT_fp16: return faiss::ScalarQuantizer::QuantizerType::QT_fp16;
+    case QuantizerType::QT_8bit_direct:
+      return faiss::ScalarQuantizer::QuantizerType::QT_8bit_direct;
+    case QuantizerType::QT_6bit: return faiss::ScalarQuantizer::QuantizerType::QT_6bit;
+    default: return (faiss::ScalarQuantizer::QuantizerType)qtype;
+  }
+}
+
+template <typename IntType = int>
+void approx_knn_ivfflat_build_index(
+  knnIndex* index, IVFParam* params, raft::distance::DistanceType metric, IntType n, IntType D)
+{
+  faiss::gpu::GpuIndexIVFFlatConfig config;
+  config.device                  = index->device;
+  faiss::MetricType faiss_metric = build_faiss_metric(metric);
+  faiss::gpu::GpuIndexIVFFlat* faiss_index =
+    new faiss::gpu::GpuIndexIVFFlat(index->gpu_res, D, params->nlist, faiss_metric, config);
+  faiss_index->setNumProbes(params->nprobe);
+  index->index = faiss_index;
+}
+
+template <typename T = float, typename IntType = int>
+void approx_knn_cuivfl_ivfflat_build_index(const raft::handle_t& handle,
+                                           knnIndex* index,
+                                           IVFParam* params,
+                                           raft::distance::DistanceType metric,
+                                           T* dataset,
+                                           IntType n,
+                                           IntType D)
+{
+  auto stream         = handle.get_stream();
+  int ratio           = 2;  // TODO: take these parameters from API
+  int niter           = 20;
+  const int dim       = D;
+  const size_t ntrain = n / ratio;
+  assert(ntrain > 0);
+
+  rmm::mr::managed_memory_resource managed_memory;
+  rmm::device_uvector<T> trainset(ntrain * dim, stream, &managed_memory);
+
+  RAFT_CUDA_TRY(cudaMemcpy2DAsync(trainset.data(),
```

Comment:
Is this assuming that the data has already been properly shuffled? Why not sample here to make sure we're getting a uniform distribution of points?

Reply:
Yes, I agree! Working on it now. I want to move the allocation of the trainset and the copying of the data inside `cuivflBuildOptimizedKmeans`, which will go into `ann_kmeans_balanced` file. Maybe I'll keep this simple copy for now, but add a TODO proper sampling there.